rustls/src/quic.rs

```diff
@@ -108,8 +164,66 @@ impl PacketKey {
                 &[],
             )),
             iv: hkdf_expand(secret, IvLen, b"quic iv", &[]),
+            suite,
         }
     }
+
+    /// Encrypt a QUIC packet
+    ///
+    /// `buf` contains the header, payload and space for the AEAD tag, with the payload starting at
+    /// `header_len` and ending at `tag_len()` before the end of `buf`.
+    pub fn encrypt(&self, packet: u64, buf: &mut [u8], header_len: usize) -> Result<(), Error> {
+        let (header, payload) = buf.split_at_mut(header_len);
+        let (payload, tag_storage) =
+            payload.split_at_mut(payload.len() - self.key.algorithm().tag_len());
+        let aad = aead::Aad::from(header);
+        let nonce = nonce_for(packet, &self.iv);
+        let tag = self
+            .key
+            .seal_in_place_separate_tag(nonce, aad, payload)
+            .map_err(|_| Error::EncryptError)?;
+        tag_storage.copy_from_slice(tag.as_ref());
+        Ok(())
+    }
+
+    /// Decrypt a QUIC packet
+    ///
+    /// Returns the length of the decrypted payload.
+    pub fn decrypt(&self, packet: u64, header: &[u8], payload: &mut [u8]) -> Result<usize, Error> {
```

Comment:
I guess I prefer not having the duplicate data between the `payload` argument and the return value.

> Alternatively, you could have separate `ciphertext_in_plaintext_out` and `tag` parameters, which would make the API very clear.

Not exactly sure what you mean here? *ring* does not seem to support passing in the tag separately for opening.

Renamed it and tried to improve the documentation along the lines you suggested.

Reply:
> Not exactly sure what you mean here? _ring_ does not seem to support passing in the tag separately for opening.

If we want to have a separate tag argument, then I'm happy to add that API to *ring*. It is needed for JWE anyway.

I'd rather design the API the way we want it, and then figure out how to implement it, rather than feel constrained by any limits of the current code.

Reply:
> I guess I prefer not having the duplicate data between the `payload` argument and the return value.

I'm not understanding this. There wouldn't be any duplication with what I'm proposing because the return slice would be borrowing the same (or a subslice) of the slice data of the input.

Reply:
There would still be two separate ways of getting to the plain text data, which makes the API harder to understand IMO. That's the reason I didn't do it, anyway. If you feel strongly about it I can change it.

Reply:
OK, I tried to implement my idea and I ran into an issue with my suggestion. Depending on how the QUIC implementation does buffer management, some of my ideas make less sense. If QUIC had defined its packet structure so that the header and tag were at the beginning then the IOVec stuff would work much better. If the QUIC implementation doesn't use IOVecs to read the header into a separate buffer from the data, then the API I've proposed is not great.

WDYT?

Reply:
I think we have one `IoSlice` per packet in Quinn, rather than having multiple slices per packet. Apart from the `IoSlice` stuff, I still kind of like separate tag opening because it puts all the slicing (header/payload/tag) in one place rather than having the header/payload split and payload/tag split handled differently.

(I was also wondering before if a similar sealing API that takes a `tag: &mut [u8]` argument rather than returning a `Tag` could avoid an additional copy.)